stable_baselines3/her/her_replay_buffer.py

```diff
@@ -0,0 +1,293 @@
+        :param env: associated gym VecEnv
+            to normalize the observations/rewards when sampling
+        :param online_sampling: Using online_sampling for HER or not.
+        :param n_sampled_goal: Number of sampled goals for replay. (offline sampling)
+        :param replay_observations: Observations of the offline replay buffer. Needed for 'RANDOM' goal strategy.
+        :return:
+        """
+        # Select which episodes to use
+        if online_sampling:
+            episode_indices = np.random.randint(0, self.n_episodes_stored, batch_size)
+            her_indices = np.arange(batch_size)[: int(self.her_ratio * batch_size)]
+            ep_length = self.episode_lengths[episode_indices]
+        else:
+            episode_length = self.episode_lengths[0]
+            episode_indices = np.array(list(range(self.n_episodes_stored)) * episode_length * n_sampled_goal)
+            her_indices = np.arange(len(episode_indices))
+            # repeat every transition index n_sampled_goals times
+            transitions_indices = np.array(list(range(episode_length)) * n_sampled_goal)
+
+        if self.goal_selection_strategy == GoalSelectionStrategy.FUTURE:
+            # restrict the sampling domain when ep_length > 1
+            # otherwise filter out the indices
+            if online_sampling:
+                her_indices = her_indices[ep_length[her_indices] > 1]
+                ep_length[her_indices] -= 1
+            else:
+                her_indices = her_indices[episode_length > 1 and transitions_indices < episode_length - 1]
+
+        if online_sampling:
+            # Select which transitions to use
+            transitions_indices = np.random.randint(ep_length)
+        # get selected transitions
+        transitions = {key: self.buffer[key][episode_indices, transitions_indices].copy() for key in self.buffer.keys()}
+
+        new_goals = self.vectorized_sample_goal(
+            episode_indices, her_indices, transitions_indices, online_sampling, replay_observations
+        )
+        transitions["desired_goal"][her_indices] = new_goals
+
+        # Convert to numpy array
+        # TODO: disable if not needed for faster computation
+        transitions["info"] = np.array(
+            [
+                self.info_buffer[episode_idx][transition_idx]
+                for episode_idx, transition_idx in zip(episode_indices, transitions_indices)
+            ]
+        )
+
+        # Vectorized computation
+        transitions["reward"][her_indices] = self.env.env_method(
+            "compute_reward",
+            transitions["next_achieved_goal"][her_indices],
+            transitions["desired_goal"][her_indices],
+            transitions["info"][her_indices],
+        )
+
+        # concatenate observation with (desired) goal
+        observations = ObsDictWrapper.convert_dict(transitions)
+        next_observations = ObsDictWrapper.convert_dict(transitions, observation_key="next_obs")
+
+        if online_sampling:
```

Comment:
probably best to not use `_sample_transitions` for the offline sampling but a new method, no?